README
NAME
    AnyEvent::DBD::Pg - AnyEvent interface to DBD::Pg's async interface

SYNOPSIS
        use AnyEvent::DBD::Pg;

        my $adb = AnyEvent::DBD::Pg->new('dbi:Pg:dbname=comments;host=db2.rambler.ru;port=6000;', comments => 'Ieigfyxbr', {
            pg_enable_utf8 => 1,
            pg_server_prepare => 0,
            quote_char => '"',
            name_sep => ".",
        }, debug => 1);

        $adb->queue_size( 4 );
        $adb->debug( 1 );

        $adb->connect;

        $adb->selectcol_arrayref("select pg_sleep( 0.1 ), 1", { Columns => [ 1 ] }, sub {
            my $rc = shift or return warn;
            my $res = shift;
            warn "Got <$adb->{qd}> = $rc / @{$res}";
            $adb->selectrow_hashref("select data,* from tx limit 2", {}, sub {
                my $rc = shift or return warn;
                warn "Got $adb->{qd} = $rc [@_]";
            });
        });

        $adb->execute("update tx set data = data;",sub {
            my $rc = shift or return warn;
            warn "Got exec: $rc";
            #my $st = shift;
            #$st->finish;
        });

        $adb->execute("select from 1",sub {
            shift or return warn;
            warn "Got $adb->{qd} = @_";
        });

        $adb->selectrow_array("select pg_sleep( 0.1 ), 2", {}, sub {
            shift or return warn;
            warn "Got $adb->{qd} = [@_]";
            $adb->selectrow_hashref("select * from tx limit 1", {}, sub {
                warn "Got $adb->{qd} = [@_]";
                $adb->execute("select * from tx", sub {
                    my $rc = shift or return warn;
                    my $st = shift;
                    while(my $row = $st->fetchrow_hashref) { warn "$row->{id}"; }
                    $st->finish;
                    exit;
                });
            });
        });
        AE::cv->recv;

METHODS
    connect()
        Establish connection to database

    selectrow_array( $query, [\%args], $cb->( $rc, ... ))
        Execute PG_ASYNC prepare, than push result of "fetchrow_array" into
        callback

    selectrow_arrayref( $query, [\%args], $cb->( $rc, \@row ))
        Execute PG_ASYNC prepare, than push result of "fetchrow_arrayref"
        into callback

    selectrow_hashref( $query, [\%args], $cb->( $rc, \%row ))
        Execute PG_ASYNC prepare, than push result of "fetchrow_hashref"
        into callback

    selectall_arrayref( $query, [\%args], $cb->( $rc, \@rows ))
        Execute PG_ASYNC prepare, than push result of "fetchall_arrayref"
        into callback

    selectall_hashref( $query, [\%args], $cb->( $rc, \@rows ))
        Execute PG_ASYNC prepare, than push result of "fetchall_hashref"
        into callback

    selectcol_arrayref( $query, { Columns => [...], ... }, $cb->( $rc,
    \@rows ))
        Execute PG_ASYNC prepare, than push result of
        "fetchall_hashref($args{Columns})" into callback

    execute( $query, [\%args], $cb->( $rc, $sth ))
        Execute PG_ASYNC prepare, than invoke callback, pushing resulting
        sth to it.

        Please, note: result already passed as first argument

AUTHOR
    Mons Anderson, "<mons@cpan.org>"

LICENSE AND COPYRIGHT
    Copyright 2010 Mons Anderson.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.



Changes
Revision history for AnyEvent-DBI-Pg

0.03_05 2012-02-08
        - freedb method
        - ping
        - remove noisy warnings about request queue

0.03_04 2011-06-17
        - Modifications to use Pool with AE::DBI

0.03_03 2011-05-10
        - No more socket fileno detection, use pg_socket! (Thanks to Ruslan Zakirov)

0.03_02 2010-12-22
        - Fix pool connect status
        - Fix selectall_arrayref Slice option

0.03    2010-11-30
        - Fix tests (remove NoWarnings)
        - Lower perl version requirement

0.02    2010-11-27
        Fix tests

0.01    2010-11-27
        First version, released on an unsuspecting world.



ex/sample.pl
	use common::sense 3;
	use lib::abs '../lib';
	use EV;
	
	use AnyEvent::DBD::Pg;
	
	my $adb = AnyEvent::DBD::Pg->new('dbi:Pg:dbname=test', user => '', {
		pg_enable_utf8 => 1,
		pg_server_prepare => 0,
		quote_char => '"',
		name_sep => ".",
	}, debug => 1);
	
	$adb->queue_size( 4 );
	$adb->debug( 1 );
	
	$adb->connect;
	
	$adb->selectcol_arrayref("select pg_sleep( 0.1 ), 1", { Columns => [ 1 ] }, sub {
		my $rc = shift or return warn;
		my $res = shift;
		warn "Got <$adb->{qd}> = $rc / @{$res}";
		$adb->selectrow_hashref("select data,* from tx limit 2", {}, sub {
			my $rc = shift or return warn;
			warn "Got $adb->{qd} = $rc [@_]";
		});
	});
	
	$adb->execute("update tx set data = data;",sub {
		my $rc = shift or return warn;
		warn "Got exec: $rc";
		#my $st = shift;
		#$st->finish;
	});
	
	$adb->execute("select from 1",sub {
		shift or return warn;
		warn "Got $adb->{qd} = @_";
	});
	
	$adb->selectrow_array("select pg_sleep( 0.1 ), 2", {}, sub {
		shift or return warn;
		warn "Got $adb->{qd} = [@_]";
		$adb->selectrow_hashref("select * from tx limit 1", {}, sub {
			warn "Got $adb->{qd} = [@_]";
			$adb->execute("select * from tx", sub {
				my $rc = shift or return warn;
				my $st = shift;
				while(my $row = $st->fetchrow_hashref) { warn "$row->{id}"; }
				$st->finish;
				exit;
			});
		});
	});
	
	AE::cv->recv;


lib/AnyEvent/DBD/Pool.pm
=for rem

$pool->txn(default => 'rollback', sub {
	my $db = shift;
	$db->do(... sub {
		
		$db->rollback;
	})
})

=cut

package AnyEvent::DBD::Pool::Tx;

use common::sense;

sub new {
	my $pkg = shift;
	my $dbh = shift;
	my $destroyer = shift;
	return bless {db => $dbh, des => $destroyer},$pkg;
}

sub commit {
	my $self = shift;
	$self->{commited} = 1;
	#warn "call commit on $self->{db}";
	$self->{db}->commit(@_ ? @_ : sub {});
}
sub rollback {
	my $self = shift;
	$self->{commited} = -1;
	#warn "call rollback on $self->{db}";
	$self->{db}->rollback(@_ ? @_ : sub {});
}

our $AUTOLOAD;
sub  AUTOLOAD {
	my $self = shift;
	my ($method) = $AUTOLOAD =~ /([^:]+)$/;
	#warn "Call $method on db";
	$self->{db}->$method(@_);
	return;
	#unshift @_, $self->{db};
	#goto &{ $self->{db}->can($method) };

}

sub DESTROY {
	my $self = shift;
	#warn "DESTROY tx";
	local $@;eval {
		$self->{des}->($self);
	1} or warn "$self/DESTROY: $@";
	return;
}

package AnyEvent::DBD::Pool;

use common::sense;
use AnyEvent::DBD::Pg;

sub new {
	my $pkg = shift;
	my $count = shift;
	my $self = bless {}, $pkg;
	my ($dsn,$user,$pass,$args,@args) = @_;
	$args ||= {};
	
	my $ctor;
	if (@args and ref $args[0] eq 'CODE') {
		$ctor = shift @args;
	} else {
		#warn "dsn = $dsn";
		my ($type) = $dsn =~ /^dbi:([^:]+):/;
		my $class = "AnyEvent::DBD::".$type;
		eval "require $class; 1" or die $@;
		$ctor = sub {
			$class->new(@_);
		};
	}
	$self->{pool_size} = $count;
	$self->{pool} = [
		map { $ctor->( $dsn,$user,$pass,$args,@args,id => $_[0] ); } 1..$count
	];
	$self->{waiting_db} = [];
	return $self;
}

sub debug {
	my $self = shift;
	$_->debug(@_) for @{ $self->{pool} };
}

sub queue_size {
	my $self = shift;
	$_->queue_size(@_) for @{ $self->{pool} };
}

sub connect {
	my $self = shift;
	my $success = 1;
	$success &&= $_->connect(@_) for @{ $self->{pool} };
	$success;
}

sub freedb {
	my $self = shift;
	my $cb = pop;
	# During this call we must prohibit free pool changing. so
	$self->{take_delay} = \ my @take;
	$self->{ret_delay} = \ my @ret;
	for (@{ $self->{pool} }) {
		$cb->($_);
	}
	delete $self->{take_delay};
	delete $self->{ret_delay};
	$self->ret(@ret);
	$self->take($_) for @take;
}

sub DESTROY {}
our $AUTOLOAD;
sub  AUTOLOAD {
	my $self = shift;
	my $cb = pop;
	my @args = @_;
	my ($method) = $AUTOLOAD =~ /([^:]+)$/;
	warn "autoloaded $AUTOLOAD";
	$self->take(sub {
		my $con = shift;
		$con->$method(@args,sub {
			$self->ret($con);
			goto &$cb;
		});
	});
}

sub txn {
	my $self = shift;
	my $cb = pop;
	my %args = (default => 'commit', @_);
	my $ix = \0;
	$self->take(sub {
		my $db = shift;
		$db->begin_work(sub {
			my $wr = AnyEvent::DBD::Pool::Tx->new($db, sub {
				my $wr = shift;
				if ($wr->{commited}) {
					$self->ret($db);
				} else {
					my $call = $args{default};
					#warn "default commit";
					$db->$call(sub{
						shift or warn;
						$self->ret($db);
					});
				}
			});
			$cb->($wr);
		});
	});
	
}

sub take {
	my $self = shift;
	my $cb = shift or die "cb required for take at @{[(caller)[1,2]]}\n";
	#warn("take wrk, left ".$#{$self->{pool}}."\n");
	if ($self->{take_delay}) {
		push @{ $self->{take_delay} }, $cb;
		return;
	}
	if (@{$self->{pool}}) {
		my $db = shift @{$self->{pool}};
		$db->{_return_to_me} = $self;
		$cb->($db);
	} else {
		push @{$self->{waiting_db}},$cb;
		if ( @{$self->{waiting_db}} > $self->{pool_size} ) {
			warn("no worker for @{[(caller 1)[1,2]]}, maybe increase pool (current penfing: @{[ 0+@{$self->{waiting_db}} ]})?");
		}
	}
	return;
}

sub ret {
	my $self = shift;
	if ($self->{ret_delay}) {
		push @{ $self->{ret_delay} }, @_;
		return;
	}
	delete $_->{_return_to_me} for @_;
	push @{ $self->{pool} }, @_;
	#warn("ret wrk, left ".$#{$self->{pool}}."; waiting=".(0+@{ $self->{waiting_db} })."\n");
	$self->take(shift @{ $self->{waiting_db} }) if @{ $self->{waiting_db} };
	return;
}


1;



lib/AnyEvent/DBD/Pg.pm
package AnyEvent::DBD::Pg;

use 5.008008; # don't use old crap without utf8
use common::sense 3;m{
	use strict;
	use warnings;
}x;
use Scalar::Util 'weaken';
use Carp;
use DBI;
use DBD::Pg ':async';
use AE 5;
use Time::HiRes 'time';

=head1 NAME

AnyEvent::DBD::Pg - AnyEvent interface to DBD::Pg's async interface

=cut

our $VERSION = '0.03_05'; $VERSION = eval($VERSION);

=head1 SYNOPSIS

	use AnyEvent::DBD::Pg;
	
	my $adb = AnyEvent::DBD::Pg->new('dbi:Pg:dbname=test', user => 'pass', {
		pg_enable_utf8 => 1,
		pg_server_prepare => 0,
		quote_char => '"',
		name_sep => ".",
	}, debug => 1);
	
	$adb->queue_size( 4 );
	$adb->debug( 1 );
	
	$adb->connect;
	
	$adb->selectcol_arrayref("select pg_sleep( 0.1 ), 1", { Columns => [ 1 ] }, sub {
		my $rc = shift or return warn;
		my $res = shift;
		warn "Got <$adb->{qd}> = $rc / @{$res}";
		$adb->selectrow_hashref("select data,* from tx limit 2", {}, sub {
			my $rc = shift or return warn;
			warn "Got $adb->{qd} = $rc [@_]";
		});
	});
	
	$adb->execute("update tx set data = data;",sub {
		my $rc = shift or return warn;
		warn "Got exec: $rc";
		#my $st = shift;
		#$st->finish;
	});
	
	$adb->execute("select from 1",sub {
		shift or return warn;
		warn "Got $adb->{qd} = @_";
	});
	
	$adb->selectrow_array("select pg_sleep( 0.1 ), 2", {}, sub {
		shift or return warn;
		warn "Got $adb->{qd} = [@_]";
		$adb->selectrow_hashref("select * from tx limit 1", {}, sub {
			warn "Got $adb->{qd} = [@_]";
			$adb->execute("select * from tx", sub {
				my $rc = shift or return warn;
				my $st = shift;
				while(my $row = $st->fetchrow_hashref) { warn "$row->{id}"; }
				$st->finish;
				exit;
			});
		});
	});
	AE::cv->recv;

=cut

sub new {
	my ($pkg,$dsn,$user,$pass,$args,@args) = @_;
	$args ||= {};
	my $self = bless {@args},$pkg;
	$self->{cnn} = [$dsn,$user,$pass,$args];
	$self->{id} = sprintf '%08x', int $self unless defined $self->{id};
	$self->{queue_size} = 2 unless defined $self->{queue_size};
	$self->{queue} = [];
	#$self->{current};
	#$self->connect;
	$self->{querynum}  = 0;
	$self->{queuetime} = 0;
	$self->{querytime} = 0;
	$self;
}

BEGIN {
	for my $method (qw( cnn queue_size debug )) {
		*$method = sub { @_ > 1 ? $_[0]->{$method} = $_[1] : $_[0]->{$method} }
	}
}

sub callcb($$@) {
	my ($cb,$err,@args) = @_;
	my $e;
	my $wa = wantarray;
	my ($rc,@rc);
	{
		local $@;
		eval {
			$@ = $err;
			$wa ? @rc = $cb->(@args) : $rc = $cb->(@args);
		1 } or $e = $@;
	}
	die $e if defined $e;
	$wa ? @rc : $rc;
}

sub _set_ka_timer {
	my $self = shift;
	$self->{ka} = AE::timer 30,30, sub {
		
	};
}

sub connect {
	my $self = shift;
	my ($dsn,$user,$pass,$args) = @{ $self->{cnn} };
	local $args->{RaiseError} = 0;
	local $args->{PrintError} = 0;
	
	if( $self->{db} = DBI->connect($dsn,$user,$pass,$args) ) {
		#warn "connect $dsn $user {@{[ %$args  ]}} successful ";
		$self->{fh} = $self->{db}->{pg_socket} + 0;
		warn "Connection to $dsn established for pid $$\n" if $self->{debug} > 2;
		$self->{pid} = $$;
		$self->{lasttry} = undef;
		$self->{gone} = undef;
		return $self->{db}->ping;
	} else {
		#warn "connect $dsn $user {@{[ %$args  ]}} failed ";
		$self->{error} = DBI->errstr;
		$self->{gone} = time unless defined $self->{gone};
		$self->{lasttry} = time;
		warn "Connection to $dsn failed: ".DBI->errstr;
		return 0;
	}
}

our %METHOD = (
	selectrow_array    => 'fetchrow_array',
	selectrow_arrayref => 'fetchrow_arrayref',
	selectrow_hashref  => 'fetchrow_hashref',
	selectall_arrayref => sub {
		my ($st,$args) = @_;
		$st->fetchall_arrayref($args->{Slice});
	}, #'fetchall_arrayref',
	selectall_hashref  => 'fetchall_hashref',
	selectcol_arrayref => sub {
		my ($st,$args) = @_;
		$st->fetchall_arrayref($args->{Columns});
	},
	execute            => sub { $_[0]; }, # just pass the $st
);

sub DESTROY {}

sub _dequeue {
	my $self = shift;
	if ($self->{db}->{pg_async_status} == 1 ) {	
		warn "Can't dequeue (), while processing query ($self->{current}[0]) by @{[ (caller)[1,2] ]}";
		if ( @{ $self->{queue} } ) {
			warn "\tHave queue $self->{queue}[0][1]  $self->{queue}[0][2]";
		}
		return;
	}
	#warn "Run dequeue with status=$self->{db}->{pg_async_status}";
	return $self->{current} = undef unless @{ $self->{queue} };
	my $next = shift @{ $self->{queue} };
	my $at = shift @$next;
	$self->{queuetime} += time - $at;
	my $method = shift @$next;
	local $self->{queuing} = 0;
	$self->$method(@$next);
}

sub ping { my $self = shift; $self->{db} && $self->{db}->ping }

sub begin_work {
	my $self = shift;
	my $cb = pop;
	$self->execute("begin",$cb);
}

sub commit {
	my $self = shift;
	my $cb = pop || sub {};
	$self->execute("commit",$cb);
}

sub rollback {
	my $self = shift;
	my $cb = pop || sub {};
	$self->execute("rollback",$cb);
}

our $AUTOLOAD;
sub  AUTOLOAD {
	my ($method) = $AUTOLOAD =~ /([^:]+)$/;
	my $self = shift;
	die sprintf qq{Can't locate autoloaded object method "%s" (%s) via package "%s" at %s line %s.\n}, $method, $AUTOLOAD, ref $self, (caller)[1,2] # '
		unless exists $METHOD{$method};
	my $fetchmethod = $METHOD{$method};
	defined $fetchmethod or croak "Method $method not implemented yet";
	ref (my $cb = pop) eq 'CODE' or croak "need callback";
	
	$self->{pid} == $$ or warn("reestablish connection to database because of pid change from $self->{pid} to $$"), $self->connect;
	$self->{db} or $self->connect or return callcb( $cb,$self->{error} );
	
	if ($self->{db}->{pg_async_status} == 1 or $self->{current} ) {
		if ( $self->{queue_size} > 0 and @{ $self->{queue} } > $self->{queue_size} - 1 ) {
			my $c = 1;
			my $counter = ++$self->{querynum};
			printf STDERR "\e[036;1m$self->{id}/Q$counter\e[0m. [\e[03${c};1m%0.4fs\e[0m] < \e[03${c};1m%s\e[0m > ".("\e[031;1mQuery run out of queue size $self->{queue_size}\e[0m")."\n", 0 , $_[0];
			return callcb( $cb, "Query $_[0] run out of queue size $self->{queue_size}" );
		} else {
			warn "Query $_[0] pushed to queue because of ".( $self->{db}->{pg_async_status} == 1 ? 'async_status' : "current=$self->{current}" )."\n" if $self->{debug} > 1;
			push @{ $self->{queue} }, [time(), $method, @_,$cb];
			return;
		}
	}
	my $query = shift;
	my $args = shift || {};
	$args->{pg_async} = PG_ASYNC;
	my $counter = ++$self->{querynum};
	warn "prepare call <$query>( @_ ), async status = ".$self->{db}->{pg_async_status} if $self->{debug} > 2;
	$self->{current} = [$query,@_];
	$self->{current_start} = time();
	
	#weaken $self;
	$self or warn("self was destroyed"),return;
	my ($st,$w,$t,$check);
	my @watchers;
	push @watchers, sub {
		$self and $st or warn("no self"), @watchers = (), return 1;
		#warn time()." Callback $self->{db}->{pg_async_status} / ".$st->pg_ready();
		my $rs;
		{
			local $SIG{__WARN__} = sub {
				warn "$self->{id}: $query.pg_ready : @_";
			};
			$rs = $self->{db}->{pg_async_status} && $st->pg_ready();
		}
		
		if ($rs) {
			undef $w;
			my $res;
			{
				local $SIG{__WARN__} = sub {
					warn "$self->{id}: $query.pg_result : @_";
				};
				$res = $st->pg_result;
			}
			my $run = time - $self->{current_start};
			$self->{querytime} += $run;
			my ($diag,$DIE);
			if ($self->{debug}) {
				$diag = $self->{current}[0];
				my @bind = @{ $self->{current} };
				shift @bind;
				$diag =~ s{\?}{ "'".shift(@bind)."'" }sge;
			} else {
				$diag = $self->{current}[0];
			}
			if (!$res) {
				$DIE = $self->{db}->errstr;
			}
			local $self->{qd} = $diag;
			if ($self->{debug}) {
				my $c = $run < 0.01 ? '2' : $run < 0.1 ? '3' : '1';
				my $x = $DIE ? '1' : '6';
				printf STDERR "\e[036;1m$self->{id}/Q$counter\e[0m. [\e[03${c};1m%0.4fs\e[0m] < \e[03${x};1m%s\e[0m > ".($DIE ? "\e[031;1m$DIE\e[0m" : '')."\n", $run , $diag;
			}
			local $self->{queuing} = @{ $self->{queue} };
			if ($res) {
				undef $@;
				#local $@;eval {
				if (ref $fetchmethod) {
					$cb->($res, $st->$fetchmethod($args));
				} else {
					$cb->($res, $st->$fetchmethod);
					
					# my @res = $st->$fetchmethod;
					# undef $st;
					# $self->_dequeue();
					# $cb->($res, @res);
				}
				#1} or do { warn "call callback failed: $@"; die $@; };
				undef $st;
				undef $self->{current};
				$self->_dequeue();
				@watchers = ();
			} else {
				#local $@ = $DIE;
				#warn "st failed: $DIE";
				$st->finish;
				callcb($cb,$DIE);
				undef $st;
				undef $self->{current};
				@watchers = ();
				$self->_dequeue();
			}
			return 1;
		}
		return 0;
		#undef $w;
	};
	if (0 and $query =~ /^\s*begin(?:_work|)\s*$/i) {
		my $rc = eval { $self->{db}->begin_work;1 };
		$rc or warn;
		
			my ($diag,$DIE);
			if (!$rc) {
				$DIE = $@;
			}
			if ($self->{debug}) {
				$diag = $self->{current}[0];
				my @bind = @{ $self->{current} };
				shift @bind;
				$diag =~ s{\?}{ "'".shift(@bind)."'" }sge;
			} else {
				$diag = $self->{current}[0];
			}
			local $self->{qd} = $diag;
			if ($self->{debug}) {
				my $c = 2;
				my $x = $DIE ? '1' : '6';
				printf STDERR "\e[036;1m$self->{id}/Q$counter\e[0m. [\e[03${c};1m%0.4fs\e[0m] < \e[03${x};1m%s\e[0m > ".($DIE ? "\e[031;1m$DIE\e[0m" : '')."\n", 0 , $diag;
			}
		
		undef $self->{current};
		if($rc) {
			$cb->( '0E0' );
		} else {
			$cb->();
		}
		$self->_dequeue if @{ $self->{queue} };
		return;
	}
	
	use Time::HiRes 'time';
	#warn time()." call query $query";
	undef $@;
	$st = eval { $self->{db}->prepare($query,$args) } # this could die with 'server closed the connection unexpectedly'
		and $st->execute(@_) 
		or return do{
			undef $st;
			@watchers = ();
			my $err = defined $@ && $@ ? "$@" : $self->{db}->errstr;
			warn "prepare/execute failed: $@/".$self->{db}->errstr. " = $err ";
			callcb($cb, $err);
			
			$self->_dequeue;
		};
	#warn time()." call query $query done";
	# At all we don't need timers for the work, but if we have some bugs, it will help us to find them
	push @watchers, AE::timer 1,1, $watchers[0];
	push @watchers, AE::io $self->{fh}, 0, $watchers[0];
	$watchers[0]() and return;
	return;
}

=head1 METHODS

=over 4

=item connect()

Establish connection to database

=item selectrow_array( $query, [\%args], $cb->( $rc, ... ))

Execute PG_ASYNC prepare, than push result of C<fetchrow_array> into callback

=item selectrow_arrayref( $query, [\%args], $cb->( $rc, \@row ))

Execute PG_ASYNC prepare, than push result of C<fetchrow_arrayref> into callback

=item selectrow_hashref( $query, [\%args], $cb->( $rc, \%row ))

Execute PG_ASYNC prepare, than push result of C<fetchrow_hashref> into callback

=item selectall_arrayref( $query, [\%args], $cb->( $rc, \@rows ))

Execute PG_ASYNC prepare, than push result of C<fetchall_arrayref> into callback

=item selectall_hashref( $query, [\%args], $cb->( $rc, \@rows ))

Execute PG_ASYNC prepare, than push result of C<fetchall_hashref> into callback

=item selectcol_arrayref( $query, { Columns => [...], ... }, $cb->( $rc, \@rows ))

Execute PG_ASYNC prepare, than push result of C<fetchall_hashref($args{Columns})> into callback

=item execute( $query, [\%args], $cb->( $rc, $sth ))

Execute PG_ASYNC prepare, than invoke callback, pushing resulting sth to it.

B<Please, note>: result already passed as first argument

=back

=head1 THANKS

=over 4

=item * Thanks to Ruslan Zakirov C<< <ruz@bestpractical.com> >> for hint about pg_socket option

=back

=head1 AUTHOR

Mons Anderson, C<< <mons@cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Mons Anderson.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

=cut

1;
